### PR TITLE
Removed UnsafeTraversableMethods - iteration 1

### PR DESCRIPTION
**Pull request by carloscaldas**
Related to #2995 

Removed `88` errors. Still missing `62` errors

This change is [Reviewable](https://reviewable.io/reviews/lbnl-ucb-sti/beam/2997)

### CODE REVIEW

File: src/main/scala/beam/agentsim/agents/ridehail/RideHailIterationHistory.scala
Comment (rajnikantsh) [inline src/main/scala/beam/agentsim/agents/ridehail/RideHailIterationHistory.scala:8]: May be typo?

Comment (carloscaldas) [reply src/main/scala/beam/agentsim/agents/ridehail/RideHailIterationHistory.scala:8]: yes. it was a typo

File: src/main/scala/beam/agentsim/agents/PersonAgent.scala
Comment (rajnikantsh) [inline src/main/scala/beam/agentsim/agents/PersonAgent.scala:638]: Observation: its reduce readability a lot and may be error prone.

Comment (carloscaldas) [reply src/main/scala/beam/agentsim/agents/PersonAgent.scala:638]: I agree it reduces the readability but is not error prone. drop never throws exception: if iterable is empty it keeps empty

File: src/main/scala/beam/agentsim/agents/choice/mode/ModeChoiceDriveIfAvailable.scala
Comment (rajnikantsh) [inline src/main/scala/beam/agentsim/agents/choice/mode/ModeChoiceDriveIfAvailable.scala:33]: Condition needs to be reviewed.

**Review by rajnikantsh — APPROVED** [review]
Except Small suggestions all LGTM

File: src/main/scala/beam/physsim/jdeqsim/EventToHourFrequency.scala
Comment (JustinPihony) [inline src/main/scala/beam/physsim/jdeqsim/EventToHourFrequency.scala:56]: Same as elsewhere, should this throw instead of resulting in an empty loop?

File: src/main/scala/beam/physsim/bprsim/Coordinator.scala
Comment (JustinPihony) [inline src/main/scala/beam/physsim/bprsim/Coordinator.scala:59]: Won't this result in another Option.get error in Scapegoat?

File: src/main/scala/beam/agentsim/agents/ridehail/RideHailModifyPassengerScheduleManager.scala
Comment (JustinPihony) [inline src/main/scala/beam/agentsim/agents/ridehail/RideHailModifyPassengerScheduleManager.scala:163]: We should double check that this is using the same logging as elsewhere, where the if isenabled is already built in (akka logger I think?)

File: src/main/scala/beam/agentsim/agents/ridehail/AsyncGreedyVehicleCentricMatching.scala
Comment (JustinPihony) [inline src/main/scala/beam/agentsim/agents/ridehail/AsyncGreedyVehicleCentricMatching.scala:87]: We should keep the comment or turn it into a method named aptly

**Review by JustinPihony — COMMENTED** [review]
Some logic changes that maybe should wait for another PR - and other suggestions

File: src/main/scala/beam/agentsim/agents/modalbehaviors/DrivesVehicle.scala
Comment (carloscaldas) [reply src/main/scala/beam/agentsim/agents/modalbehaviors/DrivesVehicle.scala:520]: It is. You can check it here:
https://github.com/LBNL-UCB-STI/beam/pull/2997/files#diff-36423ace99c365be70fd706e76182c87R458

File: src/main/scala/beam/replanning/SupplementaryTripGenerator.scala
Comment (carloscaldas) [reply src/main/scala/beam/replanning/SupplementaryTripGenerator.scala:116]: ok

File: src/main/scala/beam/physsim/bprsim/Coordinator.scala
Comment (carloscaldas) [reply src/main/scala/beam/physsim/bprsim/Coordinator.scala:59]: in this case will be solved during in the PR that fixes Options on scapegoat. At this moment I am trying to keep the same behaviour

File: src/main/scala/beam/agentsim/agents/ridehail/RideHailModifyPassengerScheduleManager.scala
Comment (carloscaldas) [reply src/main/scala/beam/agentsim/agents/ridehail/RideHailModifyPassengerScheduleManager.scala:163]: `isEnabled` is already implemented in all the logs system. But it will avoid only string objects creation. It won't avoid the calculation of `min`/ `max` that are expensive operations

File: src/main/scala/beam/replanning/SupplementaryTripGenerator.scala
Comment (carloscaldas) [reply src/main/scala/beam/replanning/SupplementaryTripGenerator.scala:80]: it can either push a bug further down or solve a bug. it is just a decision to be done

File: src/main/scala/beam/agentsim/agents/ridehail/AsyncGreedyVehicleCentricMatching.scala
Comment (carloscaldas) [reply src/main/scala/beam/agentsim/agents/ridehail/AsyncGreedyVehicleCentricMatching.scala:87]: It just changed the line number. Some github issue makes it keep showing
image

File: src/main/scala/beam/agentsim/agents/household/FastHouseholdCAVScheduling.scala
Comment (carloscaldas) [reply src/main/scala/beam/agentsim/agents/household/FastHouseholdCAVScheduling.scala:142]: issue created and TODO removed:
https://github.com/LBNL-UCB-STI/beam/issues/2999

File: src/main/scala/beam/agentsim/agents/modalbehaviors/ModeChoiceCalculator.scala
Comment (carloscaldas) [reply src/main/scala/beam/agentsim/agents/modalbehaviors/ModeChoiceCalculator.scala:118]: Issue created https://github.com/LBNL-UCB-STI/beam/issues/3000

File: src/main/scala/beam/agentsim/agents/planning/Tour.scala
Comment (carloscaldas) [reply src/main/scala/beam/agentsim/agents/planning/Tour.scala:10]: Issue created: https://github.com/LBNL-UCB-STI/beam/issues/3000

File: src/main/scala/beam/agentsim/agents/choice/mode/ModeChoiceMultinomialLogit.scala
Comment (carloscaldas) [reply src/main/scala/beam/agentsim/agents/choice/mode/ModeChoiceMultinomialLogit.scala:63]: as you can see it is a huge change. maybe better just create a new PR for those refactoring. otherwise does not finish. there is another PR just to remove 90 messages of scapegoat

File: src/main/scala/beam/agentsim/agents/PersonAgent.scala
Comment (carloscaldas) [reply src/main/scala/beam/agentsim/agents/PersonAgent.scala]: fixed typo and changed the orElse log

File: src/main/scala/beam/agentsim/agents/PersonAgent.scala
Comment (carloscaldas) [reply src/main/scala/beam/agentsim/agents/PersonAgent.scala:661]: asked them already. this is not reason to block this PR

Comment (REASY) [reply src/main/scala/beam/agentsim/agents/PersonAgent.scala:661]: @carloscaldas should we introduce safe analogues for all those unsafe methods, like [tail](https://www.scala-lang.org/api/2.12.2/scala/collection/GenTraversableLike.html#head:A) and others?

File: src/main/scala/beam/physsim/jdeqsim/EventToHourFrequency.scala
Comment (carloscaldas) [reply src/main/scala/beam/physsim/jdeqsim/EventToHourFrequency.scala:56]: if there are no maxHour then csv file won't be written. on the other hand if there is behavior does not change. In the previous code an exception is thrown. Considering this code is there for a long time I believe is working fine.
This rule is the same as replacing `last` with`drop(1)`.

Comment (carloscaldas) [reply src/main/scala/beam/physsim/jdeqsim/EventToHourFrequency.scala:56]: I think we should avoid throwing exception. is the same as replacing `tail` with `drop(1)`

File: src/main/scala/beam/agentsim/agents/choice/mode/ModeChoiceDriveIfAvailable.scala
Comment (carloscaldas) [reply src/main/scala/beam/agentsim/agents/choice/mode/ModeChoiceDriveIfAvailable.scala]: done

File: src/main/scala/beam/agentsim/agents/modalbehaviors/ModeChoiceCalculator.scala
Comment (carloscaldas) [reply src/main/scala/beam/agentsim/agents/modalbehaviors/ModeChoiceCalculator.scala:118]: this is the same tradeoff as basically all the changes

File: src/main/scala/beam/agentsim/agents/PersonAgent.scala
Comment (JustinPihony) [reply src/main/scala/beam/agentsim/agents/PersonAgent.scala:661]: Sure - we should create an issue to at least review it though

File: src/main/scala/beam/agentsim/agents/choice/mode/ModeChoiceDriveIfAvailable.scala
Comment (JustinPihony) [inline src/main/scala/beam/agentsim/agents/choice/mode/ModeChoiceDriveIfAvailable.scala:32]: This comment was lost in a change I think:

chooseRandomElement changes the logic itself. The random index throws explicitly, so now we are hiding this until later. If kept, we should remove that method altogether as it is only used in one other location anyway.

Comment (JustinPihony) [reply src/main/scala/beam/agentsim/agents/choice/mode/ModeChoiceDriveIfAvailable.scala:32]: I am catching up on all comments and now see your reply is:

> this is the same tradeoff as basically all the changes

While I agree then my latter comment comes into play - we should remove the index method altogether - or make it have the same semantics

File: src/main/scala/beam/utils/SequenceUtils.scala
Comment (JustinPihony) [inline src/main/scala/beam/utils/SequenceUtils.scala:30]: This needs to be Opt also (for consistency in naming)

File: src/main/scala/beam/physsim/jdeqsim/EventToHourFrequency.scala
Comment (JustinPihony) [reply src/main/scala/beam/physsim/jdeqsim/EventToHourFrequency.scala:56]: That is a larger debate - maybe worth bringing up at the next standup - I'm of the mindset that an exception early or at least a log for something that might result in unexpected output..but becomes hard to track or worse - isn't noticed at all.

File: src/main/scala/beam/agentsim/agents/ridehail/RideHailModifyPassengerScheduleManager.scala
Comment (JustinPihony) [reply src/main/scala/beam/agentsim/agents/ridehail/RideHailModifyPassengerScheduleManager.scala:163]: true - wasn't paying attention to those being done outside of the string

File: src/main/scala/beam/agentsim/agents/ridehail/AsyncGreedyVehicleCentricMatching.scala
Comment (JustinPihony) [reply src/main/scala/beam/agentsim/agents/ridehail/AsyncGreedyVehicleCentricMatching.scala:87]: sorry about that :)

File: src/main/scala/beam/agentsim/agents/choice/mode/ModeChoiceDriveIfAvailable.scala
Comment (carloscaldas) [reply src/main/scala/beam/agentsim/agents/choice/mode/ModeChoiceDriveIfAvailable.scala:32]: Lets the the usage of previous method:
```
class ModeChoiceUniformRandom(val beamConfig: BeamConfig) extends ModeChoiceCalculator {

  override def apply(
    alternatives: IndexedSeq[EmbodiedBeamTrip],
    attributesOfIndividual: AttributesOfIndividual,
    destinationActivity: Option[Activity],
    person: Option[Person] = None
  ): Option[EmbodiedBeamTrip] = {
    if (alternatives.nonEmpty) {
      Some(alternatives(chooseRandomAlternativeIndex(alternatives)))
    } else {
      None
    }
  }
```

**Review by JustinPihony — APPROVED** [review]

**Comment by carloscaldas** [comment]
Test now!

**Comment by REASY** [comment]
In order to sync codebase with another fork I have to revert this PR, the reason is there are too many changes in PersonAgent, RideHailManager. Once the merge process will finish, let's bring back it.

**Comment by JustinPihony** [comment]
@REASY What fork, so we can monitor - it appears it is still even with develop, so something is missing to me

**Pull request merged**
